lib/bitbucket/token_validator.go

```diff
@@ -0,0 +1,82 @@
+/*
+ * Teleport
+ * Copyright (C) 2023  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package bitbucket
+
+import (
+	"context"
+	"time"
+
+	"github.com/coreos/go-oidc"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+
+	"github.com/gravitational/teleport/lib/jwt"
+)
+
+// IDTokenValidator validates a Bitbucket issued ID Token.
+type IDTokenValidator struct {
+	// clock is used by the validator when checking expiry and issuer times of
+	// tokens. If omitted, a real clock will be used.
+	clock clockwork.Clock
+}
+
+// NewIDTokenValidator returns an initialized IDTokenValidator
+func NewIDTokenValidator(clock clockwork.Clock) *IDTokenValidator {
+	if clock == nil {
+		clock = clockwork.NewRealClock()
+	}
+
+	return &IDTokenValidator{
+		clock: clock,
+	}
+}
+
+// Validate validates a Bitbucket issued ID token.
+func (id *IDTokenValidator) Validate(
+	ctx context.Context, issuerURL, audience, token string,
+) (*IDTokenClaims, error) {
+	p, err := oidc.NewProvider(ctx, issuerURL)
+	if err != nil {
+		return nil, trace.Wrap(err, "creating oidc provider")
+	}
```

Comment:
Should we cache the provider for some time? Is there any risk of:
- DDoS-ing the bitbucket provider
- Getting rate-limited by the Bitbucket OIDC provider
- Exhausting our outbound connections (NAT saturationsx, too many go clients, ...)

How are the other join methods protecting against a join surge?

Reply:
https://github.com/gravitational/teleport/issues/49213